docs/compatibility_draft.md

```diff
@@ -0,0 +1,246 @@
+
+_Description:_
+- 1/1/23: Create bytecode_v1 from libStablehlo@v1
+- 1/7/23: Rename MyOp → TheOp, mnemonic remains MyDialect.MyOp in libStablehlo@v2
+  + No change to serialization / deserialization
+  + Requires changes to APIs, depending on API stability guarantees.
+  + `bytecode_v2` can be read in libStablehlo@v1 since mnemonic is unchanged
+
+### Migrating a Feature (Rename / Move)
+_Characteristics of a migration change:_
+- Change in dialect breaks serialization in a semantically compatible manner (IR able to be upgraded/downgraded).
+- Downgrade bytecode for 3 weeks.
+- Upgrade bytecode for 6 months.
+
+_Example:_ 
+- Rename an Operation mnemonic, `cross-replica-sum → cross_replica_sum`.
+
+_Compatibility Protocol:_
+
+<img width="832" alt="migration" src="https://user-images.githubusercontent.com/3036970/189246202-04f426f5-d12e-4eee-8b54-5c9e1396b6f7.png">
+
+_Description:_
+- 1/1/23: Create bytecode_v1 from libStablehlo@v1
+- 1/7/23: Rename op and mnemonic `MyOp → TheOp`, build libStablehlo@v2
+  + No need to version op, can upgrade from / downgrade to generic form.
+  + Implement upgrade pass on generic form for backwards compat.
+  + libStablehlo@v2 can handle `MyOp@v1` w/ upgrade or `TheOp@v2` without.
+  + Downgrade bytecode `TheOp → MyOp` for forwards compatibility.
+- 1/29/23: Bytecode bytecode_v3 from libStablehlo@v3
+  + No need for downgrade, v2 can handle `TheOp@v3` bytecode.
+  + Upgrade pass from `MyOp@v1 → MyOp@v2` maintained for 6 months.
+
+### Adding a Feature
+_Characteristics of a addition change:_
+- A new feature is introduced, with new semantics
+- No forward compatibility when feature is used
+- Warn on forward compatibility if bytecoding new feature
+
+_Example:_ Adding a new enum value or new attribute.
+
+_Compatibility Protocol:_
+
+<img width="818" alt="addition" src="https://user-images.githubusercontent.com/3036970/189246224-7d9df1cd-bec4-4a6f-b8fc-4a41e5f794f5.png">
+
+_Description:_
+- 1/1/23: Create `bytecode_v1` from libStablehlo@v1
+- 1/7/23: Add enum `<B>` which has new semantics in libStablehlo@v2
+  + No upgrade needed since `<A>` exists in both builds.
+  + Add bytecode verifier pass with `emitWarning` if `<B>` is used, warning of forwards incompatibility.
+- 1/29/23: Remove warning in libStablehlo@v3
+  + Forward and backward compatibility fully supported after 3 weeks.
+
+### Deprecating a Feature
+_Characteristics of a deprecation change:_
+- Remove a feature with backwards compatibility
+- Prevent creation of new bytecode using feature
+- Clean up after backward compatibility window
+
+_Example:_ Removing an op / attribute / enum.
+
+_Compatibility Protocol:_
+
+<img width="951" alt="deprecate" src="https://user-images.githubusercontent.com/3036970/189246238-96ef5fdd-a735-40ef-982d-c4a1b5cc114a.png">
+
+_Description:_
+- 1/1/23: Create bytecode_v1 from libStablehlo@v1
+- 1/7/23: Add verifier to prevent serialization of `<B>` in libStablehlo@v2
```

Comment:
What's a bytecode verifier? Is this referring to the pass that you wrote or that's a check performed by the serialization logic? I'd say we should do all this in once place, so I'd recommend this would go into a pass then.

Also, let's discuss a good name for this thing. I think we can improve on "bytecode verifier" because: a) it doesn't operate on bytecode, b) "verifier" already means something else in MLIR.

How about "compatibility checker"? I feel that it adequately reflects what this thing does. Then we could name the pass `--stablehlo-compatibility-checker` which is still a fairly long name, but shorter than the current one.

Reply:
Sounds good. I agree that avoiding the word "verifier", except when referring to MLIR verifiers, is a good idea.